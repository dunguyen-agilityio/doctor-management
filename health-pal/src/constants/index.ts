export * from './device'
export * from './environment'
export * from './message'
export * from './regex'
export * from './route'
export * from './specialty'
