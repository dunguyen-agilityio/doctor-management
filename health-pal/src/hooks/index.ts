export * from './use-app-loading'
