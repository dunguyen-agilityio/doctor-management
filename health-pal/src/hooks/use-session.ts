import { useAuthStore } from '@app/stores/auth'

export const useSession = () => {
  const { user, isAuthenticated, signIn, signOut, setUser } = useAuthStore()

  return { session: { user, isAuthenticated }, signIn, signOut, setUser }
}
