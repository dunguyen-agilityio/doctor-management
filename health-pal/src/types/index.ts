export * from './auth'
export * from './booking'
export * from './doctor'
export * from './favorite'
export * from './image'
export * from './modal'
export * from './route'
export * from './specialty'
export * from './strapi'
