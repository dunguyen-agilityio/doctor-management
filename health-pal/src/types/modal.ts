export type ModalRef<T = null> = { open: (params?: T) => void; close: () => void }
