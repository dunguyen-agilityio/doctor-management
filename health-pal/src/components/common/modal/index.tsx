import { useEffect } from 'react'

import { Button, Dialog, DialogContentProps, DialogProps, Unspaced } from 'tamagui'

import { WINDOW_SIZE } from '@app/constants'

import { useModal } from '@app/hooks/use-modal'

import { X } from '@icons'

import { ModalRef } from '@app/types/modal'

export interface ModalProps extends DialogContentProps, Omit<DialogProps, 'children'> {
  closeButtonShown?: boolean
  ref?: React.Ref<ModalRef>
}

const Modal = ({
  children,
  closeButtonShown,
  ref,
  open: defaultOpen,
  ...props
}: React.PropsWithChildren<ModalProps>) => {
  const [open, setOpen] = useModal(ref)

  useEffect(() => {
    if (defaultOpen) {
      setOpen(true)
    }
  }, [defaultOpen, setOpen])

  return (
    <Dialog modal open={open} onOpenChange={setOpen}>
      <Dialog.Portal>
        <Dialog.Overlay
          testID="overlay"
          backgroundColor="$shadow3"
          animateOnly={['transform', 'opacity']}
          animation={[
            'quicker',
            {
              opacity: {
                overshootClamping: true,
              },
            },
          ]}
          enterStyle={{ opacity: 0, scale: 0.95 }}
          exitStyle={{ opacity: 0 }}
        />

        <Dialog.Content
          role="dialog"
          testID="content"
          padding={0}
          width={WINDOW_SIZE.width - 48}
          minHeight={WINDOW_SIZE.height / 2}
          borderRadius="$2"
          overflow="hidden"
          shadowColor="#fff"
          shadowOffset={{ height: 0, width: 0 }}
          key="content"
          animateOnly={['transform', 'opacity']}
          animation={[
            'quicker',
            {
              opacity: {
                overshootClamping: true,
              },
            },
          ]}
          enterStyle={{ x: 0, y: -20, opacity: 0 }}
          exitStyle={{ x: 0, y: 10, opacity: 0, scale: 0.95 }}
          {...props}>
          {children}
          {closeButtonShown && (
            <Unspaced>
              <Dialog.Close asChild>
                <Button
                  position="absolute"
                  top="$md"
                  right="$md"
                  size="$5"
                  circular
                  icon={X}
                  aria-label="Close the dialog"
                  accessibilityHint="Closes the cancellation dialog without making changes"
                  role="button"
                />
              </Dialog.Close>
            </Unspaced>
          )}
        </Dialog.Content>
      </Dialog.Portal>
    </Dialog>
  )
}

export default Modal
