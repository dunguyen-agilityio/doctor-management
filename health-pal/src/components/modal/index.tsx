import { useImperativeHandle, useState } from 'react'

import { X } from '@tamagui/lucide-icons'
import { Button, Dialog, DialogContentProps, DialogProps, Unspaced } from 'tamagui'

import { WINDOW_SIZE } from '@app/constants'

import { ModalRef } from '@app/types/modal'

interface ModalProps extends DialogProps {
  closeButtonShown?: boolean
  contentProps?: DialogContentProps
  ref?: React.Ref<ModalRef>
}

const Modal = ({
  children,
  closeButtonShown,
  contentProps,
  ref,
  ...props
}: React.PropsWithChildren<ModalProps>) => {
  const [open, setOpen] = useState(false)

  useImperativeHandle(ref, () => ({
    open: () => setOpen(true),
    close: () => setOpen(false),
  }))

  return (
    <Dialog modal open={open} onOpenChange={setOpen} {...props}>
      <Dialog.Portal>
        <Dialog.Overlay
          key="overlay"
          backgroundColor="$shadow3"
          animateOnly={['transform', 'opacity']}
          animation={[
            'quicker',
            {
              opacity: {
                overshootClamping: true,
              },
            },
          ]}
          enterStyle={{ opacity: 0, scale: 0.95 }}
          exitStyle={{ opacity: 0 }}
        />

        <Dialog.Content
          padding={0}
          width={WINDOW_SIZE.width - 48}
          minHeight={WINDOW_SIZE.height / 2}
          borderRadius="$2"
          overflow="hidden"
          shadowColor="#fff"
          shadowOffset={{ height: 0, width: 0 }}
          key="content"
          animateOnly={['transform', 'opacity']}
          animation={[
            'quicker',
            {
              opacity: {
                overshootClamping: true,
              },
            },
          ]}
          enterStyle={{ x: 0, y: -20, opacity: 0 }}
          exitStyle={{ x: 0, y: 10, opacity: 0, scale: 0.95 }}
          {...contentProps}>
          {children}
          {closeButtonShown && (
            <Unspaced>
              <Dialog.Close asChild>
                <Button position="absolute" top="$md" right="$md" size="$5" circular icon={X} />
              </Dialog.Close>
            </Unspaced>
          )}
        </Dialog.Content>
      </Dialog.Portal>
    </Dialog>
  )
}

export default Modal
