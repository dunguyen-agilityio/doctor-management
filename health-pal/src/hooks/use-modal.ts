import { useImperativeHandle, useState } from 'react'

import { ModalRef } from '@/types/modal'

export const useModal = (ref?: React.Ref<ModalRef | null>) => {
  const [open, setOpen] = useState(false)

  useImperativeHandle(ref, () => ({
    open: () => setOpen(true),
    close: () => setOpen(false),
  }))

  return [open, setOpen] as const
}
