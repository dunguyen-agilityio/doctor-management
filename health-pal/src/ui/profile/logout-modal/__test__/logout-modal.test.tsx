import { act, fireEvent, render, screen } from '@utils-test'

import { router } from 'expo-router'

import { ROUTES } from '@/constants'

import * as MockModal from '@/hooks/use-modal'
import { useSession } from '@/hooks/use-session'

import { ModalRef } from '@/types/modal'

import { queryClient } from '@react-query.config'

import LogoutModal from '..'

jest.mock('@react-query.config')
jest.mock('expo-router')

// Mock useSession
jest.mock('@/hooks/use-session', () => ({
  useSession: jest.fn(),
}))

describe('LogoutModal', () => {
  const mockSignOut = jest.fn()
  const mockRef: React.RefObject<ModalRef> = {
    current: {
      open: jest.fn(),
      close: jest.fn(),
    },
  }

  let setOpenMock: jest.Mock

  beforeEach(() => {
    jest.clearAllMocks()
    ;(useSession as jest.Mock).mockReturnValue({
      signOut: mockSignOut,
    })

    setOpenMock = jest.fn()
    jest.spyOn(MockModal, 'useModal').mockImplementation((ref) => {
      if (ref && ref.current) {
        ref.current.open = mockRef.current!.open
        ref.current.close = mockRef.current!.close
      }
      return [true, setOpenMock]
    })
  })
  it('should match snapshot', () => {
    const tree = render(<LogoutModal ref={mockRef} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('closes modal when Cancel button is clicked', () => {
    render(<LogoutModal ref={mockRef} />)
    const cancelButton = screen.getByLabelText('Cancel logout')
    act(() => {
      fireEvent.press(cancelButton)
    })
    expect(mockRef.current?.close).toHaveBeenCalled()
  })

  it('handles logout when Yes, Logout button is clicked', () => {
    render(<LogoutModal ref={mockRef} />)
    const logoutButton = screen.getByLabelText('Confirm logout')
    fireEvent.press(logoutButton)
    expect(router.replace).toHaveBeenCalledWith(ROUTES.LOGIN)
    expect(queryClient.clear).toHaveBeenCalled()
    expect(mockSignOut).toHaveBeenCalled()
  })
})
