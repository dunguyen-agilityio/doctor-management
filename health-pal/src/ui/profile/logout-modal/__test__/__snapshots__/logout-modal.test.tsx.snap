// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoutModal should match snapshot 1`] = `
[
  <View
    aria-label="Notifications (F8)"
    pointerEvents="box-none"
    role="region"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 100000,
      }
    }
    tabIndex={-1}
  >
    <View
      focusable={false}
      pointerEvents="box-none"
      style={
        {
          "flexDirection": "column",
          "maxWidth": "100%",
          "position": "absolute",
        }
      }
    />
  </View>,
  <View
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 105001,
      }
    }
  >
    <View
      collapsable={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "rgba(0, 0, 0, 0.4)",
          "bottom": 0,
          "flexDirection": "column",
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": [],
          "zIndex": 99999,
        }
      }
      testID="sheet-overlay"
    />
    <View
      onLayout={[Function]}
      style={
        {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "pointerEvents": "none",
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      animation={null}
      collapsable={false}
      disableClassName={true}
      onLayout={[Function]}
      style={
        [
          {
            "height": undefined,
            "minHeight": undefined,
            "opacity": 1,
            "position": "absolute",
            "width": "100%",
            "zIndex": 100000,
          },
          {
            "transform": [
              {
                "translateY": 10000.1,
              },
            ],
          },
        ]
      }
    >
      <View
        accessibilityHint="Dialog to confirm logging out of your account"
        aria-label="Logout confirmation"
        pointerEvents="auto"
        role="dialog"
        style={
          {
            "backgroundColor": "hsla(0, 0%, 100%, 1)",
            "borderTopLeftRadius": 34,
            "borderTopRightRadius": 34,
            "flex": 0,
            "flexDirection": "column",
            "gap": 24,
            "height": 200,
            "maxHeight": "100%",
            "overflow": "hidden",
            "paddingBottom": 24,
            "paddingLeft": 24,
            "paddingRight": 24,
            "paddingTop": 24,
            "width": "100%",
          }
        }
        testID="sheet-modal"
      >
        <RCTScrollView
          onMoveShouldSetResponder={[Function]}
          onResponderRelease={[Function]}
          onScroll={[Function]}
          onStartShouldSetResponder={[Function]}
          scrollEnabled={true}
          scrollEventThrottle={8}
          style={
            {
              "flex": 1,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "backgroundColor": "white",
                  "flexDirection": "column",
                  "width": "100%",
                }
              }
            >
              <Text
                aria-label="Logout"
                style={
                  {
                    "color": "#1c2a3a",
                    "fontFamily": "Inter_700Bold",
                    "fontSize": 20,
                    "lineHeight": 30,
                    "textAlign": "center",
                  }
                }
                suppressHighlighting={true}
              >
                Logout
              </Text>
              <View
                style={[]}
              />
              <Text
                aria-label="Are you sure you want to log out?"
                style={
                  {
                    "color": "#6b7280",
                    "fontFamily": "Inter_600SemiBold",
                    "fontSize": 16,
                    "lineHeight": 24,
                    "textAlign": "center",
                  }
                }
                suppressHighlighting={true}
              >
                Are you sure you want to log out?
              </Text>
            </View>
            <View
              style={
                {
                  "backgroundColor": "white",
                  "flexDirection": "row",
                  "gap": 16,
                  "paddingTop": 16,
                }
              }
            >
              <View
                accessibilityHint="Cancels the logout action and closes the dialog"
                aria-label="Cancel logout"
                focusVisibleStyle={{}}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                role="button"
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#e5e7eb",
                    "borderBottomColor": "transparent",
                    "borderBottomLeftRadius": 60,
                    "borderBottomRightRadius": 60,
                    "borderBottomWidth": 1,
                    "borderLeftColor": "transparent",
                    "borderLeftWidth": 1,
                    "borderRightColor": "transparent",
                    "borderRightWidth": 1,
                    "borderStyle": "solid",
                    "borderTopColor": "transparent",
                    "borderTopLeftRadius": 60,
                    "borderTopRightRadius": 60,
                    "borderTopWidth": 1,
                    "color": "#1c2a3a",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "nowrap",
                    "fontWeight": "700",
                    "height": 48,
                    "justifyContent": "center",
                    "opacity": 1,
                    "paddingBottom": 12,
                    "paddingLeft": 20,
                    "paddingRight": 20,
                    "paddingTop": 12,
                  }
                }
              >
                <Text
                  lineBreakMode="clip"
                  numberOfLines={1}
                  style={
                    {
                      "color": "#1c2a3a",
                      "flexGrow": 0,
                      "flexShrink": 1,
                      "fontFamily": "Inter_700Bold",
                      "fontSize": 14,
                      "lineHeight": 21,
                      "userSelect": "none",
                    }
                  }
                  suppressHighlighting={true}
                >
                  Cancel
                </Text>
              </View>
              <View
                accessibilityHint="Logs out of your account and closes the dialog"
                aria-label="Confirm logout"
                focusVisibleStyle={{}}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                role="button"
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#1c2a3a",
                    "borderBottomColor": "transparent",
                    "borderBottomLeftRadius": 60,
                    "borderBottomRightRadius": 60,
                    "borderBottomWidth": 1,
                    "borderLeftColor": "transparent",
                    "borderLeftWidth": 1,
                    "borderRightColor": "transparent",
                    "borderRightWidth": 1,
                    "borderStyle": "solid",
                    "borderTopColor": "transparent",
                    "borderTopLeftRadius": 60,
                    "borderTopRightRadius": 60,
                    "borderTopWidth": 1,
                    "color": "#ffffff",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "nowrap",
                    "fontWeight": "700",
                    "height": 48,
                    "justifyContent": "center",
                    "opacity": 1,
                    "paddingBottom": 12,
                    "paddingLeft": 20,
                    "paddingRight": 20,
                    "paddingTop": 12,
                  }
                }
              >
                <Text
                  lineBreakMode="clip"
                  numberOfLines={1}
                  style={
                    {
                      "color": "#ffffff",
                      "flexGrow": 0,
                      "flexShrink": 1,
                      "fontFamily": "Inter_700Bold",
                      "fontSize": 14,
                      "lineHeight": 21,
                      "userSelect": "none",
                    }
                  }
                  suppressHighlighting={true}
                >
                  Yes, Logout
                </Text>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        accessibilityHint="Dialog to confirm logging out of your account"
        aria-label="Logout confirmation"
        role="dialog"
        style={
          {
            "backgroundColor": "hsla(0, 0%, 100%, 1)",
            "borderBottomLeftRadius": 0,
            "borderBottomRightRadius": 0,
            "borderBottomWidth": 0,
            "borderLeftWidth": 0,
            "borderRightWidth": 0,
            "borderStyle": "solid",
            "borderTopLeftRadius": 0,
            "borderTopRightRadius": 0,
            "borderTopWidth": 0,
            "bottom": "-100%",
            "flex": 1,
            "flexDirection": "column",
            "gap": 24,
            "height": 0,
            "left": 0,
            "maxHeight": "100%",
            "overflow": "hidden",
            "paddingBottom": 24,
            "paddingLeft": 24,
            "paddingRight": 24,
            "paddingTop": 24,
            "position": "absolute",
            "right": 0,
            "shadowColor": undefined,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": "100%",
            "zIndex": -1,
          }
        }
        testID="sheet-modal"
      />
    </View>
  </View>,
]
`;
